TestTestgresCommon::test_child_pids[remote] has problem

https://github.com/postgrespro/testgres/actions/runs/29373027538/job/87226314560#step:10:5829

[testgres--test_logs--20260714_231504-altlinux_11-py3_xx_xx-pg17_xx-id29373027538.zip](https://github.com/user-attachments/files/30031451/testgres--test_logs--20260714_231504-altlinux_11-py3_xx_xx-pg17_xx-id29373027538.zip)

```
__________________ TestTestgresCommon.test_child_pids[remote] __________________
[gw3] linux -- Python 3.12.7 /tmp/testgres_venv/bin/python3

self = <tests.test_testgres_common.TestTestgresCommon object at 0x7f70e193e5d0>
node_svc = <tests.helpers.global_data.PostgresNodeService object at 0x7f70e2992930>

    def test_child_pids(self, node_svc: PostgresNodeService):
        assert isinstance(node_svc, PostgresNodeService)
    
        master_processes = [
            ProcessType.AutovacuumLauncher,
            ProcessType.BackgroundWriter,
            ProcessType.Checkpointer,
            ProcessType.StatsCollector,
            ProcessType.WalSender,
            ProcessType.WalWriter,
        ]
    
        postgresVersion = get_pg_version2(node_svc.os_ops)
    
        if __class__.helper__pg_version_ge(postgresVersion, '10'):
            master_processes.append(ProcessType.LogicalReplicationLauncher)
    
        if __class__.helper__pg_version_ge(postgresVersion, '14'):
            master_processes.remove(ProcessType.StatsCollector)
    
        repl_processes = [
            ProcessType.Startup,
            ProcessType.WalReceiver,
        ]
    
        def LOCAL__test_auxiliary_pids(
            node: PostgresNode,
            expectedTypes: typing.List[ProcessType]
        ) -> typing.List[ProcessType]:
            # returns list of the absence processes
            assert node is not None
            assert type(node) is PostgresNode
            assert expectedTypes is not None
            assert type(expectedTypes) is list
    
            pids = node.auxiliary_pids
            assert pids is not None
            assert type(pids) is dict
    
            result: typing.List[ProcessType] = list()
            for ptype in expectedTypes:
                if ptype not in pids:
                    result.append(ptype)
        run_r      = (1, '', '')
        self       = <testgres.operations.remote_ops.RemoteOperations object at 0x7f70e2538770>
        shell      = True
        ssh_cmd    = ['ssh',
 '-i',
 '/home/test/testgres/id_ed25519_test_ci',
 '-p',
 '22',
 'test@machine2-altlinux_11-py3_xx_xx-pg17_xx',
 'trap \'\' HUP;export LANG="C.UTF-8";ps -p 116327 -o cmd --no-headers']
        stderr     = None
        stdin      = None
        stdout     = None
        text       = False
        timeout    = None
        verbose    = False
        wait_exit  = False
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd = ['ps', '-p', '116327', '-o', 'cmd', '--no-headers'], exit_code = 1
msg_arg = b'', error = '', out = ''

    @staticmethod
    def UtilityExitedWithNonZeroCode(cmd, exit_code, msg_arg, error, out) -> typing.NoReturn:
        assert type(exit_code) is int
    
        msg_arg_s = __class__._TranslateDataIntoString(msg_arg)
        assert type(msg_arg_s) is str
    
        msg_arg_s = msg_arg_s.strip()
        if msg_arg_s == "":
            msg_arg_s = "#no_error_message"
    
        message = "Utility exited with non-zero code (" + str(exit_code) + "). Error: `" + msg_arg_s + "`"
>       raise ExecUtilException(
            message=message,
            command=cmd,
            exit_code=exit_code,
            out=out,
            error=error)
E       testgres.operations.exceptions.ExecUtilException: Utility exited with non-zero code (1). Error: `#no_error_message`
E       Command: ps -p 116327 -o cmd --no-headers
E       Exit code: 1

__class__  = <class 'testgres.operations.raise_error.RaiseError'>
cmd        = ['ps', '-p', '116327', '-o', 'cmd', '--no-headers']
error      = ''
exit_code  = 1
message    = 'Utility exited with non-zero code (1). Error: `#no_error_message`'
msg_arg    = b''
msg_arg_s  = '#no_error_message'
out        = ''

/tmp/testgres_venv/lib64/python3/site-packages/testgres/operations/raise_error.py:20: ExecUtilException
------------------------------ Captured log call -------------------------------
DEBUG    root:port_manager__generic2.py:157 [port manager] Port 21650 is reserved.
DEBUG    root:port_manager__generic2.py:157 [port manager] Port 26241 is reserved.
INFO     root:test_testgres_common.py:1023 Test pids of [testgres-e41e539d-6ccb-4242-8458-632268f7ee31] node. Attempt #1.
INFO     root:test_testgres_common.py:1035 Bingo!
INFO     root:test_testgres_common.py:1023 Test pids of [testgres-0362ff36-a9c1-4db7-bf05-c120ea39447b] node. Attempt #1.
INFO     root:test_testgres_common.py:1035 Bingo!
DEBUG    root:port_manager__generic2.py:157 [port manager] Port 26241 is released.
DEBUG    root:port_manager__generic2.py:157 [port manager] Port 21650 is released.
=============================== warnings summary ===============================
tests/test_testgres_remote.py::TestTestgresRemote::test_init__LANG_С
  /tmp/testgres_venv/lib64/python3/site-packages/_pytest/fixtures.py:1313: PytestRemovedIn10Warning: Class-scoped fixture defined as instance method is deprecated.
  Instance attributes set in this fixture will NOT be visible to test methods,
  as each test gets a new instance while the fixture runs only once per class.
  Use @classmethod decorator and set attributes on cls instead.
  See https://docs.pytest.org/en/stable/deprecations.html#class-scoped-fixture-as-instance-method
    fixturefunc = resolve_fixture_function(fixturedef, request)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_testgres_common.py::TestTestgresCommon::test_child_pids[remote] - testgres.operations.exceptions.ExecUtilException: Utility exited with non-zero code (1). Error: `#no_error_message`
Command: ps -p 116327 -o cmd --no-headers
Exit code: 1
====== 1 failed, 1292 passed, 3 skipped, 1 warning in 5455.59s (1:30:55) =======

```

--- a/Dockerfile--altlinux_10.tmpl
+++ b/Dockerfile--altlinux_10.tmpl
@@ -129,11 +129,6 @@ if [ ! -f /home/test/.ssh/id_rsa ]; then \
     chmod 600 /home/test/.ssh/authorized_keys; \
 fi; \
 ls -la /home/test/.ssh/; \
-su -c \"ssh-keyscan -H localhost >> /home/test/.ssh/known_hosts\" test; \
-su -c \"ssh-keyscan -H 127.0.0.1 >> /home/test/.ssh/known_hosts\" test; \
-if [ -n \"${TEST_CFG__REMOTE_HOST:-}\" ]; then \
-    su -c \"ssh-keyscan -H ${TEST_CFG__REMOTE_HOST} >> /home/test/.ssh/known_hosts\" test; \
-fi; \
 if [ -n \"${TEST_CFG__REMOTE_SSH_KEY:-}\" ]; then \
     cp \"${TEST_CFG__REMOTE_SSH_KEY}\" \"${TEST_CFG__REMOTE_SSH_KEY}_ci\"; \
     export TEST_CFG__REMOTE_SSH_KEY=\"${TEST_CFG__REMOTE_SSH_KEY}_ci\"; \

--- a/Dockerfile--altlinux_11.tmpl
+++ b/Dockerfile--altlinux_11.tmpl
@@ -130,11 +130,6 @@ if [ ! -f /home/test/.ssh/id_rsa ]; then \
     chmod 600 /home/test/.ssh/authorized_keys; \
 fi; \
 ls -la /home/test/.ssh/; \
-su -c \"ssh-keyscan -H localhost >> /home/test/.ssh/known_hosts\" test; \
-su -c \"ssh-keyscan -H 127.0.0.1 >> /home/test/.ssh/known_hosts\" test; \
-if [ -n \"${TEST_CFG__REMOTE_HOST:-}\" ]; then \
-    su -c \"ssh-keyscan -H ${TEST_CFG__REMOTE_HOST} >> /home/test/.ssh/known_hosts\" test; \
-fi; \
 if [ -n \"${TEST_CFG__REMOTE_SSH_KEY:-}\" ]; then \
     cp \"${TEST_CFG__REMOTE_SSH_KEY}\" \"${TEST_CFG__REMOTE_SSH_KEY}_ci\"; \
     export TEST_CFG__REMOTE_SSH_KEY=\"${TEST_CFG__REMOTE_SSH_KEY}_ci\"; \

--- a/src/impl/platforms/linux/internal_platform_utils.py
+++ b/src/impl/platforms/linux/internal_platform_utils.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 from .. import internal_platform_utils as base
 from ... import internal_utils
+from ....raise_error import RaiseError
 
 from testgres.operations.os_ops import OsOperations
 from testgres.operations.exceptions import ExecUtilException
@@ -417,21 +418,12 @@ class InternalPlatformUtils(base.InternalPlatformUtils):
         assert type(data_dir) is str
         assert type(failures) is list
 
-        err_msg = "FindPostmater did {} attempts and has not gotten a stable result. List of failures:\n"
+        method_name = "InternalPlatformUtils::FindPostmaster(bin_dir={!r}, data_dir={!r})".format(
+            bin_dir,
+            data_dir,
+        )
 
-        n = 0
-        sep = ""
-        for e in failures:
-            assert isinstance(e, Exception)
-
-            n += 1
-            err_msg += sep
-            err_msg += "Failure #{}. Exception ({}):\n{}".format(
-                n,
-                type(e).__name__,
-                str(e),
-            )
-            sep = "\n"
-            continue
-
-        raise RuntimeError(err_msg)
+        RaiseError.function_did_multiple_attempts_without_stable_result(
+            method_name,
+            failures,
+        )

--- a/src/node.py
+++ b/src/node.py
@@ -78,6 +78,7 @@ from .exceptions import \
 from .port_manager import PortManager
 from .impl.port_manager__this_host import PortManager__ThisHost
 from .impl.port_manager__generic2 import PortManager__Generic2
+from .impl import internal_utils
 
 from .logger import TestgresLogger
 
@@ -157,6 +158,8 @@ class ProcessProxy(object):
 class PostgresNode(object):
     # a max number of node start attempts
     _C_MAX_START_ATEMPTS = 5
+
+    _C_MAX_GET_CHILDREN_ATTEMPTS = 5
 
     _C_PM_PID__IS_NOT_DETECTED = -1
 
@@ -459,10 +462,65 @@ class PostgresNode(object):
         assert type(pid) is int
         assert isinstance(self._os_ops, OsOperations)
 
-        # get a list of postmaster's children
-        children = self._os_ops.get_process_children(pid)
+        C_MAX_ATTEMPT_COUNT = __class__. _C_MAX_GET_CHILDREN_ATTEMPTS
+        assert type(C_MAX_ATTEMPT_COUNT) is int
+        assert C_MAX_ATTEMPT_COUNT > 0
 
-        return [ProcessProxy(p) for p in children]
+        failures: typing.List[Exception] = []
+
+        nAttempt = 0
+
+        while True:
+            assert nAttempt < C_MAX_ATTEMPT_COUNT
+
+            nAttempt += 1
+
+            # get a list of postmaster's children
+            children = self._os_ops.get_process_children(pid)
+            assert type(children) is list
+
+            result: typing.List[ProcessProxy] = []
+
+            for p in children:
+                assert hasattr(p, "pid")
+                try:
+                    proxy = ProcessProxy(p)  # raise
+                except Exception as e:
+                    internal_utils.send_log_debug(
+                        "Failed to process a node child process [pid: {}]. Exception ({}): {}".format(
+                            p.pid,
+                            type(e).__name__,
+                            e,
+                        )
+                    )
+                    failures.append(e)
+                    break
+
+                assert type(proxy) is ProcessProxy
+                result.append(proxy)
+                continue
+
+            if len(result) == len(children):
+                return result
+
+            assert len(result) < len(children)
+
+            if nAttempt < C_MAX_ATTEMPT_COUNT:
+                time.sleep(0.05)
+                continue
+            break
+
+        assert nAttempt == C_MAX_ATTEMPT_COUNT
+        assert len(failures) == C_MAX_ATTEMPT_COUNT
+
+        method_name = "PostgresNode::_get_child_processes(pid={!r})".format(
+            pid,
+        )
+
+        RaiseError.function_did_multiple_attempts_without_stable_result(
+            method_name,
+            failures,
+        )
 
     @property
     def source_walsender(self):

--- a/src/node.py
+++ b/src/node.py
@@ -451,10 +451,10 @@ class PostgresNode(object):
         if x.pid is None:
             assert x.node_status != NodeStatus.Running
             RaiseError.node_err__cant_enumerate_child_processes(
-                x.node_status
+                x.node_status,
             )
 
-        assert x.node_status == NodeStatus.Running
+        assert x.node_status != NodeStatus.Stopped
         assert type(x.pid) is int
         return self._get_child_processes(x.pid)
 

--- a/src/raise_error.py
+++ b/src/raise_error.py
@@ -6,14 +6,14 @@ import typing
 
 class RaiseError:
     @staticmethod
-    def pg_ctl_returns_an_empty_string(_params):
+    def pg_ctl_returns_an_empty_string(_params) -> typing.NoReturn:
         errLines = []
         errLines.append("Utility pg_ctl returns an empty string.")
         errLines.append("Command line is {0}".format(_params))
         raise RuntimeError("\n".join(errLines))
 
     @staticmethod
-    def pg_ctl_returns_an_unexpected_string(out, _params):
+    def pg_ctl_returns_an_unexpected_string(out, _params) -> typing.NoReturn:
         errLines = []
         errLines.append("Utility pg_ctl returns an unexpected string:")
         errLines.append(out)
@@ -22,7 +22,7 @@ class RaiseError:
         raise RuntimeError("\n".join(errLines))
 
     @staticmethod
-    def pg_ctl_returns_a_zero_pid(out, _params):
+    def pg_ctl_returns_a_zero_pid(out, _params) -> typing.NoReturn:
         errLines = []
         errLines.append("Utility pg_ctl returns a zero pid. Output string is:")
         errLines.append(out)
@@ -33,7 +33,7 @@ class RaiseError:
     @staticmethod
     def node_err__cant_enumerate_child_processes(
         node_status: NodeStatus
-    ):
+    ) -> typing.NoReturn:
         assert type(node_status) is NodeStatus
 
         msg = "Can't enumerate node child processes. {}.".format(
@@ -48,7 +48,7 @@ class RaiseError:
     @staticmethod
     def node_err__cant_kill(
         node_status: NodeStatus
-    ):
+    ) -> typing.NoReturn:
         assert type(node_status) is NodeStatus
 
         msg = "Can't kill server process. {}.".format(

--- a/src/raise_error.py
+++ b/src/raise_error.py
@@ -61,6 +61,38 @@ class RaiseError:
         raise InvalidOperationException(msg)
 
     @staticmethod
+    def function_did_multiple_attempts_without_stable_result(
+        function_name: str,
+        failures: typing.List[Exception],
+    ) -> typing.NoReturn:
+        assert type(function_name) is str
+        assert type(failures) is list
+
+        err_msg = "{} did {} attempts and has not gotten a stable result.".format(
+            function_name,
+            len(failures),
+        )
+
+        err_msg += " List of failures:\n"
+
+        n = 0
+        sep = ""
+        for e in failures:
+            assert isinstance(e, Exception)
+
+            n += 1
+            err_msg += sep
+            err_msg += "Failure #{}. Exception ({}):\n{}".format(
+                n,
+                type(e).__name__,
+                str(e),
+            )
+            sep = "\n"
+            continue
+
+        raise InvalidOperationException(err_msg)
+
+    @staticmethod
     def _map_node_status_to_reason(
         node_status: NodeStatus,
         node_pid: typing.Optional[int],

